File: Next-Greater-Element-II.py
class Solution:
    def nextGreaterElements(self, nums):
        n = len(nums)
        res = [-1] * n  
        stack = []

        for i in range(2 * n - 1, -1, -1):
            while stack and stack[-1] <= nums[i % n]:
                stack.pop()

            if i < n and stack:
                res[i % n] = stack[-1]

            stack.append(nums[i % n])

        return res

File: Next-Greater-Element-II.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        stack<int> s;
        int n = nums.size();
        vector<int> res(n, -1);  
        
        for(int i = 2 * n - 1; i >= 0; i--){
            while(!s.empty() && s.top() <= nums[i % n]){
                s.pop();
            }
            
            if (i < n && !s.empty()){
                res[i % n] = s.top();
            }
            
            s.push(nums[i % n]);
        }
        return res;
    }
};
